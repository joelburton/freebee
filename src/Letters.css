.Letters {
  position: relative;
  width: 350px;
  height: 360px;
}

.Letters .Letter:nth-child(1) {
  top: 123px;
  left: 125px;
}

.Letters .Letter:nth-child(2) {
  top: 10px;
  left: 125px;
}

.Letters .Letter:nth-child(3) {
  top: 65px;
  left: 20px;
}

.Letters .Letter:nth-child(4) {
  top: 180px;
  left: 20px;
}

.Letters .Letter:nth-child(5) {
  top: 235px;
  left: 125px;
}

.Letters .Letter:nth-child(6) {
  top: 180px;
  left: 230px;
}

.Letters .Letter:nth-child(7) {
  top: 65px;
  left: 230px;
}

.Letters-randomize {
  top: 5px;
  left: 5px;
  position: relative;
  font-size: 200%;
  color: gray;
  border: none;
  background-color: white;
  cursor: crosshair;
}
