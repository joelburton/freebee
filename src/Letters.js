import { useState } from "react";
import { shuffle } from "lodash";

import Letter from "./Letter";

import "./Letters.css";

/** Letters of game board.
 *
 * Props:
 * - letters: string of non-center letters ("rstlna")
 * - center: center letter (eg "e")
 *
 * State:
 * - randomLetters: list of letters as currently arranged
 *
 * Game -> Letters -> { Letter, Letter, ... }
 */

function Letters({ letters, center }) {
  const [randomLetters, setRandomLetters] = useState([...letters]);
  console.info("* Letters", letters, center, randomLetters);

  function randomize() {
    setRandomLetters((letters) => shuffle(letters));
  }

  return (
    <div className="Letters">
      <Letter letter={center} isCenter />
      {randomLetters.map((letter) => (
        <Letter key={letter} letter={letter} />
      ))}
      <button className="Letters-randomize" onClick={randomize}>â†»</button>
    </div>
  );
}

export default Letters;
