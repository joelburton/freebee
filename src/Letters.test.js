import lodash from "lodash";

lodash.shuffle = jest.fn();

import { fireEvent, render } from "@testing-library/react";
import Letters from "./Letters";

function getLetterN(container, n) {
  return container.querySelector(`.Letter:nth-child(${n})`);
}

it("shows letter", function () {
  const { getByText } = render(<Letters letters="ABCDEF" center="Z" />);

  const a = getByText("A");
  const b = getByText("B");
  const c = getByText("C");
  const d = getByText("D");
  const e = getByText("E");
  const f = getByText("F");
  const z = getByText("Z");
  expect(a).toBeInTheDocument();
  expect(b).toBeInTheDocument();
  expect(c).toBeInTheDocument();
  expect(d).toBeInTheDocument();
  expect(e).toBeInTheDocument();
  expect(f).toBeInTheDocument();
  expect(z).toBeInTheDocument();
});

it("randomizes", function () {
  lodash.shuffle.mockReturnValueOnce(["A", "B", "C", "D", "F", "E"]);

  const { container } = render(<Letters letters="ABCDEF" center="Z" />);
  expect(getLetterN(container, 6)).toHaveTextContent("E");
  expect(getLetterN(container, 7)).toHaveTextContent("F");

  fireEvent.click(container.querySelector(".Letters-randomize"));
  expect(getLetterN(container, 6)).toHaveTextContent("F");
  expect(getLetterN(container, 7)).toHaveTextContent("E");
});

it("matches snapshot", function () {
  const { container } = render(<Letters letters="ABCDEF" center="Z" />);
  expect(container).toMatchSnapshot();
});
