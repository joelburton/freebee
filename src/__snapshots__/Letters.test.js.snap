// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="Letters"
  >
    <div
      class="Letter Letter-isCenter"
    >
      <div
        class="Letter-hex"
      >
        ⬢
      </div>
      <span>
        Z
      </span>
    </div>
    <div
      class="Letter "
    >
      <div
        class="Letter-hex"
      >
        ⬢
      </div>
      <span>
        A
      </span>
    </div>
    <div
      class="Letter "
    >
      <div
        class="Letter-hex"
      >
        ⬢
      </div>
      <span>
        B
      </span>
    </div>
    <div
      class="Letter "
    >
      <div
        class="Letter-hex"
      >
        ⬢
      </div>
      <span>
        C
      </span>
    </div>
    <div
      class="Letter "
    >
      <div
        class="Letter-hex"
      >
        ⬢
      </div>
      <span>
        D
      </span>
    </div>
    <div
      class="Letter "
    >
      <div
        class="Letter-hex"
      >
        ⬢
      </div>
      <span>
        E
      </span>
    </div>
    <div
      class="Letter "
    >
      <div
        class="Letter-hex"
      >
        ⬢
      </div>
      <span>
        F
      </span>
    </div>
    <button
      class="Letters-randomize"
    >
      ↻
    </button>
  </div>
</div>
`;
